blanderbit/vue-ads-table-tree
# Extend vue-ads-table-tree table functionality.

# Definition of Done:

- [X] Added possibility to hide/show table columns.
- [X] Added possibility to fix table head.
- [X] The `vue-ads-pagination` component supports a new custom prop: `with-input`.
- [X] Added possibility to apply a new build of `vue-ads-pagination` package.
- [X] Added possibility to set the column width.
- [X] Added possibility to make a table columns resizable.
- [X] Fix issue with incorrect sorting when some column key contains `Date` string value.

# Internal changes:

- Added new table properties that can change its behavior:
  - `has-top-slot` - Render top slot with default content (filter input).
  - `fixed-table-head` - Fix table head while scrolling.
  - `manage-table-properties` - Show property manager. Allows you to control the visibility of columns.
  - `pagination-with-page-input` - Indicates whether to display an input field (`true`) or buttons (`false`) to navigate to a specific page.
  - `columns-resizable` - Make table columns resizable.
- Added a new styling class called `manageProperties`, that allows to style the column selection manager.
Keep in mind that you cannot hide all columns. If you try, your last visible column will be disabled
in the dropdown (means you can not hide it). You always will have at least one visible column.
- Table column is clickable this is done to avoid clicking on sort icon.
- The `vue-ads-pagination` package functionality has been extended.
Added support for the ability to display either the navigation page buttons or the input field,
which allows you to quickly navigate to the desired page.
The build for this functionality was taken from `develop` branch of the forked repository.  
The corresponding PR for this functionality you can find [here](https://github.com/arnedesmedt/vue-ads-pagination/pull/27).
To be able to install a new build of the `vue-ads-pagination` package, in root directory of the `vue-ads-table-tree` project 
a new folder has been created called `deps`.
This folder contains the updated `vue-ads-pagination` package build in the corresponding folder.
Added a new npm script called `postinstall`, which runs after modules are installed and applies the new package build.
- Added a `utils` file that contains utilities functions.
- Added a new `utils` function called `isValidDate`, that checks if passed value is `Date`.

# Testing Done:

It was tested on latest FF, Chrome, and Edge.

1. Run application locally.
2. Click on sliders icon on top right above table.
3. You should see table columns in dropdown.
4. Try to hide one column.
5. You should not see this column in the table.
6. If you try to hide all columns, you will still have one visible column.
7. This one visible column will be disabled in the table properties dropdown (means you can not toggle it).
